transports/quic/tests/smoke.rs

```diff
@@ -286,3 +290,118 @@ async fn dial_failure() -> Result<()> {
+            return TestResult::discard();
+        }
+
+        let mut pool = futures::executor::LocalPool::default();
+        let mut data = vec![0; 4096 * 10];
+        rand::thread_rng().fill_bytes(&mut data);
+        let mut listeners = vec![];
+
+        // Spawn the listener nodes.
+        for _ in 0..number_listeners {
+            let mut listener = block_on(create_swarm(true)).unwrap();
+            Swarm::listen_on(&mut listener, "/ip4/127.0.0.1/udp/0/quic".parse().unwrap()).unwrap();
+
+            // Wait to listen on address.
+            let addr = match block_on(listener.next()) {
+                Some(SwarmEvent::NewListenAddr { address, .. }) => address,
+                e => panic!("{:?}", e),
+            };
+
+            listeners.push((*listener.local_peer_id(), addr));
+
+            pool.spawner()
+                .spawn_obj(
+                    async move {
+                        loop {
+                            match listener.next().await {
+                                Some(SwarmEvent::ConnectionEstablished { .. }) => {}
+                                Some(SwarmEvent::IncomingConnection { .. }) => {}
+                                Some(SwarmEvent::Behaviour(RequestResponseEvent::Message {
+                                    message:
+                                        RequestResponseMessage::Request {
+                                            request: Ping(ping),
+                                            channel,
+                                            ..
+                                        },
+                                    ..
+                                })) => {
+                                    listener
+                                        .behaviour_mut()
+                                        .send_response(channel, Pong(ping))
+                                        .unwrap();
+                                }
+                                Some(SwarmEvent::Behaviour(
+                                    RequestResponseEvent::ResponseSent { .. },
+                                )) => {}
+                                Some(SwarmEvent::ConnectionClosed { .. }) => {}
+                                e => panic!("{:?}", e),
+                            }
+                        }
+                    }
+                    .boxed()
+                    .into(),
+                )
+                .unwrap();
+        }
+
+        let mut dialer = block_on(create_swarm(true)).unwrap();
+
+        // For each listener node start `number_streams` requests.
+        for (listener_peer_id, listener_addr) in listeners {
+            dialer
+                .behaviour_mut()
+                .add_address(&listener_peer_id, listener_addr);
+
+            for _ in 0..number_streams {
+                dialer
+                    .behaviour_mut()
+                    .send_request(&listener_peer_id, Ping(data.clone()));
```

Comment:
So I got a reproducible error with number_listeners=1, number_streams=2. And I debugged it a little bit.

* If number_streams=1, then:

<details><summary>CLICK ME</summary>
```
2022-02-21T22:19:16.471310Z  INFO prop{number_listeners=1 number_streams=1}:create_swarm{keylog=true}: smoke: peer_id=PeerId("12D3KooWFc25seNNmzjU31RcdQ8kybNnNmrY8tQvZ5EBsjereKbs")
2022-02-21T22:19:16.474056Z  INFO prop{number_listeners=1 number_streams=1}:create_swarm{keylog=true}: smoke: peer_id=PeerId("12D3KooWS78yehcDhQtA76jvekkFqnU7ZM517YTWzNBaqiDgApcv")
2022-02-21T22:19:16.474506Z  INFO prop{number_listeners=1 number_streams=1}:try_send_request: libp2p_request_response: peer not connected
2022-02-21T22:19:16.474852Z  INFO prop{number_listeners=1 number_streams=1}: smoke: dialer Dialing
2022-02-21T22:19:16.477580Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:19:16.480826Z  INFO prop{number_listeners=1 number_streams=1}: smoke: listener IncomingConnection
2022-02-21T22:19:16.481264Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:19:16.481663Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:19:16.484304Z  INFO poll: libp2p_quic::upgrade: event=Ready(Connected)
2022-02-21T22:19:16.485011Z  INFO prop{number_listeners=1 number_streams=1}:inject_connection_established{peer=PeerId("12D3KooWFc25seNNmzjU31RcdQ8kybNnNmrY8tQvZ5EBsjereKbs") conn=ConnectionId(0) endpoint=Dialer { address: "/ip4/127.0.0.1/udp/50447/quic/p2p/12D3KooWFc25seNNmzjU31RcdQ8kybNnNmrY8tQvZ5EBsjereKbs" } _errors=Some([])}: libp2p_request_response: return=()
```
</details>
And test passes.<br /><br />

* If number_streams=2, then:

<details><summary>CLICK ME</summary>
```
2022-02-21T22:20:55.196925Z  INFO prop{number_listeners=1 number_streams=2}:create_swarm{keylog=true}: smoke: peer_id=PeerId("12D3KooWSLuYD79Masc7vWnjBSjEub2wZfcvKa8nGBXvFVgbz4U8")
2022-02-21T22:20:55.199519Z  INFO prop{number_listeners=1 number_streams=2}:create_swarm{keylog=true}: smoke: peer_id=PeerId("12D3KooWF8yMitRvwRFNVkDYoMSQYFPR6JZPRnrfmhEaA7EmMhK9")
2022-02-21T22:20:55.199958Z  INFO prop{number_listeners=1 number_streams=2}:try_send_request: libp2p_request_response: peer not connected
2022-02-21T22:20:55.200070Z  INFO prop{number_listeners=1 number_streams=2}:try_send_request: libp2p_request_response: peer not connected
2022-02-21T22:20:55.200410Z  INFO prop{number_listeners=1 number_streams=2}: smoke: dialer Dialing
2022-02-21T22:20:55.200576Z  INFO prop{number_listeners=1 number_streams=2}:inject_dial_failure{peer=Some(PeerId("12D3KooWSLuYD79Masc7vWnjBSjEub2wZfcvKa8nGBXvFVgbz4U8")) _dial_error=DialPeerConditionFalse(Disconnected)}: libp2p_request_response: return=()
2022-02-21T22:20:55.200656Z  INFO prop{number_listeners=1 number_streams=2}: smoke: dialer e=Some(Behaviour(OutboundFailure { peer: PeerId("12D3KooWSLuYD79Masc7vWnjBSjEub2wZfcvKa8nGBXvFVgbz4U8"), request_id: RequestId(1), error: DialFailure }))
2022-02-21T22:20:55.200764Z  INFO prop{number_listeners=1 number_streams=2}: smoke: dialer e=Some(Behaviour(OutboundFailure { peer: PeerId("12D3KooWSLuYD79Masc7vWnjBSjEub2wZfcvKa8nGBXvFVgbz4U8"), request_id: RequestId(2), error: DialFailure }))
2022-02-21T22:20:55.202648Z  INFO poll: libp2p_quic::upgrade: event=Pending
```
</details>

Dialing stops, test fails.<br /><br />

* But if number_streams=2 and I change `RequestResponse::send_request` `.condition(dial_opts::PeerCondition::Disconnected)`  to `PeerCondition::Always`, I got:

<details><summary>CLICK ME</summary>
```
2022-02-21T22:16:51.250703Z  INFO prop{number_listeners=1 number_streams=2}:try_send_request: libp2p_request_response: peer not connected
2022-02-21T22:16:51.250826Z  INFO prop{number_listeners=1 number_streams=2}:try_send_request: libp2p_request_response: peer not connected
2022-02-21T22:16:51.251160Z  INFO prop{number_listeners=1 number_streams=2}: smoke: dialer Dialing
2022-02-21T22:16:51.251429Z  INFO prop{number_listeners=1 number_streams=2}: smoke: dialer Dialing
2022-02-21T22:16:51.253403Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:16:51.254043Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:16:51.256479Z  INFO prop{number_listeners=1 number_streams=2}: smoke: listener IncomingConnection
2022-02-21T22:16:51.256921Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:16:51.258330Z  INFO prop{number_listeners=1 number_streams=2}: smoke: listener IncomingConnection
2022-02-21T22:16:51.258388Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:16:51.258747Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:16:51.259084Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:16:51.261180Z  INFO poll: libp2p_quic::upgrade: event=Ready(Connected)
2022-02-21T22:16:51.262058Z  INFO prop{number_listeners=1 number_streams=2}:inject_connection_established{peer=PeerId("12D3KooWFMA9Zo7BJinx6Jm95N4rJcYDZkjbdQnj9iJXJGPQPLTE") conn=ConnectionId(0) endpoint=Dialer { address: "/ip4/127.0.0.1/udp/32856/quic/p2p/12D3KooWFMA9Zo7BJinx6Jm95N4rJcYDZkjbdQnj9iJXJGPQPLTE" } _errors=Some([])}: libp2p_request_response: return=()
2022-02-21T22:16:51.262315Z  INFO prop{number_listeners=1 number_streams=2}:inject_connected{peer=PeerId("12D3KooWFMA9Zo7BJinx6Jm95N4rJcYDZkjbdQnj9iJXJGPQPLTE")}:try_send_request: libp2p_request_response: peer connected
2022-02-21T22:16:51.262422Z  INFO prop{number_listeners=1 number_streams=2}:inject_connected{peer=PeerId("12D3KooWFMA9Zo7BJinx6Jm95N4rJcYDZkjbdQnj9iJXJGPQPLTE")}:try_send_request: libp2p_request_response: peer connected
2022-02-21T22:16:51.262499Z  INFO prop{number_listeners=1 number_streams=2}:inject_connected{peer=PeerId("12D3KooWFMA9Zo7BJinx6Jm95N4rJcYDZkjbdQnj9iJXJGPQPLTE")}: libp2p_request_response: return=()
```
</details>

And test passes, even with number_listeners=5, number_streams=12.

Reply:
I am not able to make the test succeed either. I hope it is the test that is buggy, I suspect it is the implementation itself.